File: services/text-postprocess/rule_regression.py
#!/usr/bin/env python3
"""hot-rule 候选规则的真实数据回归：拿 3807 条真实 ASR 输出逐条跑，
统计改动量、抽样人工核对，并检查误伤。

用法:
    python3 rule_regression.py [--dataset DIR] [--show K]
"""
import argparse
import json
import os
import re
import sys

# ---- 候选新规则（顺序即应用顺序；与 hot-rule.txt 保持一致）----
# 注意：本脚本用 Python re，且刻意不用 \b（Python 的 \w 含中文，语义与 JS 不同）
NEW_RULES = [
    ("Tailscale 统一写法",
     r"[Tt][Aa][Ii][Ll]\s*[Ss][Cc][Aa][Ll][Ee]",
     "Tailscale"),
    ("tcale -> Tailscale",
     r"(?<![A-Za-z])tcale(?![A-Za-z])",
     "Tailscale"),
    ("github 统一写法",
     r"[Gg][Ii][Tt]\s*[Hh][Uu][Bb](?![A-Za-z])",
     "GitHub"),
    ("webstick -> VibeStick",
     r"(?<![A-Za-z])webstick(?![A-Za-z])",
     "VibeStick"),
    ("agx -> AGX",
     r"(?<![A-Za-z])agx(?![A-Za-z])",
     "AGX"),
    ("中英之间补空格（中文→拉丁）",
     r"([\u3400-\u4dbf\u4e00-\u9fff])([A-Za-z])",
     r"\1 \2"),
    ("中英之间补空格（拉丁→中文）",
     r"([A-Za-z])([\u3400-\u4dbf\u4e00-\u9fff])",
     r"\1 \2"),
]


def load(ds):
    out = []
    p = os.path.join(ds, "metadata.jsonl")
    for line in open(p, encoding="utf-8"):
        try:
            d = json.loads(line)
        except Exception:
            continue
        t = d.get("final_text") or d.get("text") or ""
        if t.strip():
            out.append((d.get("created_at", ""), t))
    return out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--dataset", default=os.path.expanduser("~/Documents/CapsWriter-Voice-Dataset"))
    ap.add_argument("--show", type=int, default=6)
    args = ap.parse_args()

    recs = load(args.dataset)
    total = len(recs)
    print(f"样本 {total} 条  —  hot-rule 候选规则回归")
    print("=" * 74)

    changed = []
    per_rule = {name: 0 for name, _, _ in NEW_RULES}
    for ts, t in recs:
        cur = t
        for name, pat, rep in NEW_RULES:
            nxt = re.sub(pat, rep, cur)
            if nxt != cur:
                per_rule[name] += 1
                cur = nxt
        if cur != t:
            changed.append((ts, t, cur))

    print("\n【逐条规则命中条目数】")
    for name, _, _ in NEW_RULES:
        print(f"  {name:36s} {per_rule[name]:5d}")

    print(f"\n【汇总】改动条目 {len(changed)} / {total}  ({len(changed)/total*100:.1f}%)")

    print(f"\n【改动实拍（最新 {args.show} 条）】")
    for ts, before, after in sorted(changed)[-args.show:]:
        print(f"  [{ts[:16]}]")
        print(f"    - {before[:110]}")
        print(f"    + {after[:110]}")

    # ---- 误伤检查 ----
    print("\n【误伤检查】")
    # 1) 改动只应发生在：空格插入 / 指定专名替换。逐条 diff 到字符级，看有没有别的字符被动
    def strip_all(s):
        return re.sub(r"[\s]", "", s)

    suspicious = []
    for ts, before, after in changed:
        b, a = strip_all(before), strip_all(after)
        if len(b) != len(a):
            suspicious.append(("长度变化", before, after))
            continue
        # 逐字符比较（去空格后应完全一致，除了专名大小写替换）
        diffs = [(x, y) for x, y in zip(b, a) if x != y]
        # 允许的差异：大小写转换，或 Tailscale/VibeStick/AGX 这类替换
        for x, y in diffs:
            if x.lower() == y.lower():
                continue
            suspicious.append(("非大小写差异", before, after))
            break
    if not suspicious:
        print("  ✓ 无：所有改动都是插入空格或大小写/专名规范化，没有增删汉字与字母")
    else:
        print(f"  ! 可疑 {len(suspicious)} 条（需人工核对）：")
        seen = set()
        for kind, before, after in suspicious:
            key = after[:30]
            if key in seen:
                continue
            seen.add(key)
            print(f"    [{kind}] - {before[:90]}")
            print(f"               + {after[:90]}")
            if len(seen) >= 6:
                break

    # 2) 空格重复 / 空格紧贴标点
    bad_space = [a for _, _, a in changed if re.search(r"  |\s[，。！？、；：]|[，。！？、；：]\s", a)]
    print(f"  连续空格或空格紧贴标点: {len(bad_space)}")
    for s in bad_space[:3]:
        print(f"    {s[:90]}")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: services/text-postprocess/rule_regression.js
#!/usr/bin/env node
/**
 * 用真实的 hotRuleReplace 模块 + 真实的 3807 条录音回归，验证新规则在 JS 正则
 * 语义下的实际效果。Python 分析与 JS 在 \b / lookbehind 上有语义差异，必须实测。
 *
 * 用法: node services/text-postprocess/rule_regression.js [--show K]
 */
const fs = require("fs");
const os = require("os");
const path = require("path");
const { HotRuleReplacer } = require("../../src/helpers/hotRuleReplace");

const SHOW = (() => {
  const i = process.argv.indexOf("--show");
  return i > -1 ? Number(process.argv[i + 1]) : 8;
})();

const ROOT = path.join(__dirname, "..", "..");
const RULE_FILE = path.join(ROOT, "assets", "hot-rule.txt");
const META = path.join(os.homedir(), "Documents", "CapsWriter-Voice-Dataset", "metadata.jsonl");

const replacer = new HotRuleReplacer({ filePath: RULE_FILE, fs });
const count = replacer.load(fs.readFileSync(RULE_FILE, "utf8"));
console.log(`规则条数: ${count}`);

const recs = [];
for (const line of fs.readFileSync(META, "utf8").split("\n")) {
  if (!line.trim()) continue;
  try {
    const d = JSON.parse(line);
    const t = d.final_text || d.text || "";
    if (t.trim()) recs.push([d.created_at || "", t]);
  } catch {
    /* 跳过坏行 */
  }
}
console.log(`样本: ${recs.length} 条\n`);

const changed = [];
let errCount = 0;
for (const [ts, text] of recs) {
  const out = replacer.apply(text);
  if (out.error) errCount += 1;
  if (out.text !== text) changed.push([ts, text, out.text]);
}

console.log(`改动条目: ${changed.length} / ${recs.length}  (${((changed.length / recs.length) * 100).toFixed(1)}%)`);
console.log(`规则运行时报错: ${errCount}\n`);

console.log(`--- 改动实拍（最新 ${SHOW} 条）---`);
for (const [ts, before, after] of changed.slice(-SHOW)) {
  console.log(`  [${String(ts).slice(0, 16)}]`);
  console.log(`    - ${before.slice(0, 110)}`);
  console.log(`    + ${after.slice(0, 110)}`);
}

// ---- 误伤检查：去空白后，字符序列只允许「大小写变化」，不允许增删 ----
const strip = (s) => s.replace(/\s/g, "");
const bad = [];
for (const [, before, after] of changed) {
  const b = strip(before);
  const a = strip(after);
  if (b.length !== a.length) {
    // 长度变化只能来自 tcale→Tailscale / webstick→VibeStick 这类专名归一
    if (!/Tailscale|VibeStick/.test(after)) bad.push(["长度变化", before, after]);
    continue;
  }
  for (let i = 0; i < b.length; i += 1) {
    if (b[i] !== a[i] && b[i].toLowerCase() !== a[i].toLowerCase()) {
      bad.push(["非大小写差异", before, after]);
      break;
    }
  }
}
console.log(`\n--- 误伤检查 ---`);
if (!bad.length) {
  console.log("  OK 所有改动仅为「插入空格」或「专名归一」，无汉字/字母增删");
} else {
  console.log(`  ! 可疑 ${bad.length} 条:`);
  for (const [kind, before, after] of bad.slice(0, 6)) {
    console.log(`    [${kind}] - ${before.slice(0, 90)}`);
    console.log(`               + ${after.slice(0, 90)}`);
  }
}

// 二次幂等：对已处理结果再跑一次，应当完全不变
let notIdempotent = 0;
for (const [, , after] of changed) {
  if (replacer.apply(after).text !== after) notIdempotent += 1;
}
console.log(`  幂等性: ${notIdempotent === 0 ? "OK 二次运行零变化" : `! ${notIdempotent} 条二次运行仍在变`}`);

// 路径/URL 保护检查（agx 规则最容易误伤）
const pathCases = [
  "capswriter-agx-client 仓库",
  "~/github/capswriter-agx-client",
  "https://github.com/redyuan43/capswriter-agx-client",
  "GitHubDesktop 客户端",
  "v1.0.22 版本",
  "第2个 192.168.31.134",
];
console.log(`\n--- 边界用例 ---`);
for (const c of pathCases) {
  const out = replacer.apply(c).text;
  console.log(`  ${out === c ? "不动 " : "改动 "} ${c}${out === c ? "" : `  →  ${out}`}`);
}
